app/uk/gov/hmrc/breathingspaceifproxy/connector/ConnectorHelper.scala

```diff
@@ -18,44 +18,87 @@ package uk.gov.hmrc.breathingspaceifproxy.connector
-      .withHeaders(response.headers.toList.map(header => (header._1, header._2.mkString(""))): _*)
-      .as(MimeTypes.JSON)
+
+    def composeHappyResponse(response: HttpResponse) = Future.successful {
+      Status(response.status)(response.body)
+        .withHeaders(
+          (Header.CorrelationId, headerSet.correlationId.value),
+          (HeaderNames.CONTENT_TYPE, MimeTypes.JSON)
+        )
+        .as(MimeTypes.JSON)
+    }
+
+    response.status match {
+      case OK => composeHappyResponse(response)
+
+      case CREATED => composeHappyResponse(response)
+
+      case NOT_FOUND =>
+        ErrorResponse(headerSet.correlationId.value.some, NOT_FOUND, NonEmptyChain(Error(RESOURCE_NOT_FOUND, "".some))).value
+
+      case _ =>
+        ErrorResponse(
+          headerSet.correlationId.value.some,
+          s"Unexpected response status '${response.status}' returned while calling ${url.value}."
+        ).value
+    }
   }
 
-  def logException(implicit url: Url, hc: HeaderCarrier): PartialFunction[Throwable, Future[Result]] = {
+  def logException(implicit url: Url, headerSet: RequiredHeaderSet): PartialFunction[Throwable, Future[Result]] = {
     case httpException: HttpException =>
       val reasonToLog = s"HTTP error(${httpException.responseCode}) while calling ${url.value}."
-      ErrorResponse(retrieveCorrelationId, httpException.responseCode, reasonToLog, httpException).value
+      ErrorResponse(headerSet.correlationId.value.some, httpException.responseCode, reasonToLog, httpException).value
 
     case throwable: Throwable =>
       val reasonToLog = s"Exception caught while calling ${url.value}."
-      ErrorResponse(retrieveCorrelationId, HttpStatus.INTERNAL_SERVER_ERROR, reasonToLog, throwable).value
+      ErrorResponse(headerSet.correlationId.value.some, HttpStatus.INTERNAL_SERVER_ERROR, reasonToLog, throwable).value
   }
 
-  def logResponse(response: HttpResponse)(implicit url: Url, hc: HeaderCarrier): Unit =
+  def logResponse(response: HttpResponse)(implicit url: Url, headerSet: RequiredHeaderSet): Unit =
     response.status match {
       case status if is2xx(status) =>
-        logger.debug(s"Status($status) for request ${request(retrieveCorrelationId)}to ${url.value}")
+        logger.debug(
+          s"Status($status) for request with ${ifCorrelationIdHeaderName}(${headerSet.correlationId.value}) to ${url.value}"
+        )
 
       case status =>
-        logger.error(s"ERROR($status) for request ${request(retrieveCorrelationId)}to ${url.value}")
+        logger.error(
+          s"ERROR($status) for request ${ifCorrelationIdHeaderName}(${headerSet.correlationId.value}) to ${url.value}"
+        )
         logger.debug(s"... with Body: ${response.body}")
     }
 
-  private def request(correlationId: Option[String]): String =
-    correlationId.fold("")(value => s"with ${Header.CorrelationId}($value) ")
+  implicit def hc(implicit headerSet: RequiredHeaderSet): HeaderCarrier = {
```

Comment:
In this implementation you are removing any potential additional header. They are not harming in any case and in future if there are new headers you have to change the code. The Authorization header for instance would come for free.
In addition in the next iterations maybe ETMP will have new headers. You will need then to discriminate in the `composeResponseFromIF` function here in the ConnectorHelper (which I want in any case to mantain even when we'll be parsing the response) to select the right set of headers to send.

Reply:
We should definitely not be allowing any other headers through from PEGA to IF, I think it is good security practice to only allow through the bare minimum (you can't really state that they are not harming as you never know). The authorization header is for PEGA to authorize with us, it is not the same authorization header that we will need to send to IF.

Reply:
For ETMP we will have a new helper method `composeResponseFromEMTP` though surely?